c++/src/kj/compat/http.h

```diff
@@ -767,6 +769,45 @@ struct HttpServerSettings {
   // request so that it can pipeline the next one. We'll give them a grace period defined by the
   // above two values -- if they hit either one, we'll close the socket, but if the request
   // completes, we'll let the connection stay open to handle more requests.
+
+  kj::Maybe<HttpServerErrorHandler&> errorHandler = nullptr;
+  // Customize how client protocol errors and service application exceptions are handled by the
+  // HttpServer. If null, HttpServerErrorHandler's default implementation will be used.
+};
+
+class HttpServerErrorHandler {
+public:
+  virtual kj::Promise<void> handleClientProtocolError(kj::StringPtr message,
+      kj::HttpHeaderTable& headerTable, kj::HttpService::Response& response);
```

Comment:
I think the header table shouldn't be passed as a method parameter. The normal idiom is to pass it into the constructor. E.g. HttpService::request() doesn't pass the header table; the HttpService is expected to have received it at constructor time.

Reply:
I agree, but then we'll have to make these all pure virtual, with no default implementation. If we want to allow users to call a default implementation, we could provide free functions which take a header table as a parameter, perhaps?

Reply:
Could the default implementations use a default empty header table? I don't think there's actually any requirement to use the same header table as other responses. In fact, the HttpHeaderTable passed to HttpService is only meant to be used when parsing requests; I think the application can technically use whatever table it wants for responses.

Reply:
Great idea, that solves it.